File: rag_memory.py
"""
RAG Memory — Idea store · Repo knowledge · Build history · Debug memory · Prob. weights
"""
from __future__ import annotations
import json
import time
from dataclasses import dataclass, field, asdict
from pathlib import Path
from typing import Any


@dataclass
class MemoryEntry:
    kind: str          # "idea" | "repo" | "build" | "debug" | "prob_weight"
    key: str
    content: Any
    ts: float = field(default_factory=time.time)
    tags: list[str] = field(default_factory=list)


class RAGMemory:
    """
    Simple file-backed vector-like store.
    In production swap _search() for a real embedding/vector DB (pgvector, Chroma …).
    """

    def __init__(self, path: str = ".rag_memory.json"):
        self._path = Path(path)
        self._store: list[MemoryEntry] = []
        self._load()

    # ── persistence ─────────────────────────────────────────────────────────
    def _load(self):
        if self._path.exists():
            raw = json.loads(self._path.read_text())
            self._store = [MemoryEntry(**r) for r in raw]

    def _save(self):
        self._path.write_text(json.dumps([asdict(e) for e in self._store], indent=2))

    # ── write ────────────────────────────────────────────────────────────────
    def store(self, kind: str, key: str, content: Any, tags: list[str] | None = None):
        # Upsert by key
        self._store = [e for e in self._store if e.key != key]
        self._store.append(MemoryEntry(kind=kind, key=key, content=content, tags=tags or []))
        self._save()

    # ── read ─────────────────────────────────────────────────────────────────
    def get(self, key: str) -> MemoryEntry | None:
        return next((e for e in self._store if e.key == key), None)

    def search(self, query: str, kind: str | None = None, top_k: int = 5) -> list[MemoryEntry]:
        """Keyword overlap search (replace with embedding search in production)."""
        q_tokens = set(query.lower().split())
        scored = []
        for e in self._store:
            if kind and e.kind != kind:
                continue
            text = json.dumps(e.content).lower()
            overlap = sum(1 for t in q_tokens if t in text)
            if overlap:
                scored.append((overlap, e))
        scored.sort(key=lambda x: -x[0])
        return [e for _, e in scored[:top_k]]

    # ── specialised helpers ──────────────────────────────────────────────────
    def store_idea(self, idea_id: str, data: dict):
        self.store("idea", f"idea:{idea_id}", data, tags=["idea"])

    def store_repo(self, repo_full_name: str, data: dict):
        self.store("repo", f"repo:{repo_full_name}", data, tags=["repo"])

    def store_build(self, build_id: str, data: dict):
        self.store("build", f"build:{build_id}", data, tags=["build"])

    def store_debug(self, debug_id: str, data: dict):
        self.store("debug", f"debug:{debug_id}", data, tags=["debug"])

    def store_prob_weights(self, weights: dict):
        self.store("prob_weight", "prob_weights:global", weights, tags=["prob"])

    def get_prob_weights(self) -> dict:
        entry = self.get("prob_weights:global")
        return entry.content if entry else {}

    def recall_context(self, query: str, top_k: int = 5) -> list[dict]:
        hits = self.search(query, top_k=top_k)
        return [{"kind": h.kind, "key": h.key, "content": h.content} for h in hits]

    def summary(self) -> dict:
        from collections import Counter
        counts = Counter(e.kind for e in self._store)
        return {"total_entries": len(self._store), "by_kind": dict(counts)}

File: src/lib/factory/core/rag-memory.ts
/**
 * RAG Memory — Persistent store for ideas, repos, builds, debug logs, probability weights
 * Ported from rag_memory.py — adapted for Next.js with file-backed + Prisma storage
 */
import fs from 'fs'
import path from 'path'

export interface MemoryEntry {
  kind: 'idea' | 'repo' | 'build' | 'debug' | 'prob_weight'
  key: string
  content: any
  ts: number
  tags: string[]
}

const MEMORY_DIR = path.join(process.cwd(), '.rag_memory')

export class RAGMemory {
  private entries: MemoryEntry[] = []
  private filePath: string

  constructor(memoryPath = '.rag_memory.json') {
    this.filePath = path.join(MEMORY_DIR, memoryPath)
    this.load()
  }

  private load() {
    try {
      if (fs.existsSync(this.filePath)) {
        const raw = fs.readFileSync(this.filePath, 'utf-8')
        this.entries = JSON.parse(raw)
      }
    } catch {
      this.entries = []
    }
  }

  private save() {
    try {
      fs.mkdirSync(path.dirname(this.filePath), { recursive: true })
      fs.writeFileSync(this.filePath, JSON.stringify(this.entries, null, 2))
    } catch (err) {
      console.error('[RAGMemory] save error:', err)
    }
  }

  store(kind: MemoryEntry['kind'], key: string, content: any, tags: string[] = []) {
    this.entries = this.entries.filter(e => e.key !== key)
    this.entries.push({ kind, key, content, ts: Date.now() / 1000, tags })
    this.save()
  }

  get(key: string): MemoryEntry | null {
    return this.entries.find(e => e.key === key) || null
  }

  search(query: string, kind?: string, topK = 5): MemoryEntry[] {
    const qTokens = new Set(query.toLowerCase().split(' '))
    const scored: [number, MemoryEntry][] = []

    for (const entry of this.entries) {
      if (kind && entry.kind !== kind) continue
      const text = JSON.stringify(entry.content).toLowerCase()
      let overlap = 0
      for (const token of qTokens) {
        if (text.includes(token)) overlap++
      }
      if (overlap > 0) scored.push([overlap, entry])
    }

    scored.sort((a, b) => b[0] - a[0])
    return scored.slice(0, topK).map(([, e]) => e)
  }

  // Specialized helpers
  storeIdea(ideaId: string, data: any) {
    this.store('idea', `idea:${ideaId}`, data, ['idea'])
  }

  storeRepo(repoFullName: string, data: any) {
    this.store('repo', `repo:${repoFullName}`, data, ['repo'])
  }

  storeBuild(buildId: string, data: any) {
    this.store('build', `build:${buildId}`, data, ['build'])
  }

  storeDebug(debugId: string, data: any) {
    this.store('debug', `debug:${debugId}`, data, ['debug'])
  }

  storeProbWeights(weights: Record<string, number>) {
    this.store('prob_weight', 'prob_weights:global', weights, ['prob'])
  }

  getProbWeights(): Record<string, number> {
    const entry = this.get('prob_weights:global')
    return entry?.content || {}
  }

  recallContext(query: string, topK = 5): any[] {
    const hits = this.search(query, undefined, topK)
    return hits.map(h => ({ kind: h.kind, key: h.key, content: h.content }))
  }

  summary(): { totalEntries: number; byKind: Record<string, number> } {
    const byKind: Record<string, number> = {}
    for (const e of this.entries) {
      byKind[e.kind] = (byKind[e.kind] || 0) + 1
    }
    return { totalEntries: this.entries.length, byKind }
  }
}

// Singleton
let memoryInstance: RAGMemory | null = null
export function getMemory(): RAGMemory {
  if (!memoryInstance) {
    memoryInstance = new RAGMemory()
  }
  return memoryInstance
}
